src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2312 @@
+                redundant_regions(f, stage_num, dims[d].var,
+                                  bounds, prods, values_computed);
+        conc_overlaps.push_back(conc_reg);
+    }
+    return conc_overlaps;
+}
+
+map<string, Box>
+DependenceAnalysis::regions_required(Function f,
+                                     const DimBounds &pure_bounds,
+                                     const set<string> &prods,
+                                     bool values_computed) {
+    map<string, Box> regions;
+    int num_stages = f.updates().size() + 1;
+    for (int s = 0; s < num_stages; s++) {
+
+        DimBounds bounds = get_stage_bounds(f, s, pure_bounds);
+        map<string, Box> stage_regions =
+                regions_required(f, s, bounds, prods, values_computed);
+
+        for (auto &reg: stage_regions) {
+            // Merge region with an existing region for the function
+            if (regions.find(reg.first) == regions.end())
+                regions[reg.first] = reg.second;
+            else
+                merge_boxes(regions[reg.first], reg.second);
+        }
+    }
+    return regions;
+}
+
+map<string, Box>
+DependenceAnalysis::regions_required(Function f, int stage_num,
+                                     const DimBounds &bounds,
+                                     const set<string> &prods,
+                                     bool values_computed) {
+
+    map<string, Box> regions;
+    // Add the query function and its region to the queue
+    deque<pair<FStage, DimBounds>> f_queue;
+    FStage start(f, stage_num);
+    f_queue.push_back(make_pair(start, bounds));
+
+    // Recursively compute the regions required
+    while(!f_queue.empty()) {
+        FStage s = f_queue.front().first;
+        DimBounds curr_bounds = f_queue.front().second;
+
+        Definition def = get_stage_definition(s.func, s.stage_num);
+        Scope<Interval> curr_scope;
+
+        const vector<Dim> &dims = def.schedule().dims();
+        for (int d = 0; d < (int)dims.size() - 1; d++) {
+            string var_name = dims[d].var;
+            internal_assert(curr_bounds.find(var_name) != curr_bounds.end());
+
+            Interval simple_bounds =
+                    Interval(simplify(curr_bounds.at(dims[d].var).min),
+                             simplify(curr_bounds.at(dims[d].var).max));
+            curr_scope.push(var_name, simple_bounds);
+        }
+
+        for (auto &val: def.values()) {
+
+            map<string, Box> curr_regions =
+                    boxes_required(val, curr_scope, func_val_bounds);
+
+            Box left_reg;
+            for (const Expr &arg: def.args()) {
+                map<string, Box> arg_regions =
+                        boxes_required(arg, curr_scope, func_val_bounds);
+
+                // Merge the regions with the regions found while looking at
+                // the values
+                for (auto &reg: arg_regions) {
+                    if (curr_regions.find(reg.first) == curr_regions.end())
```

Comment:
here and elsewhere, Halide always prefers braces for if-then-else, even for single-line clauses